docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,373 @@
+  to apply the horizontal scale.
+
+## Discussion and Limitations
+Overall, multidimensional autoscaling is a job which cannot be accomplished efficiently with the one-dimensional
+autoscalers currently available in the K8s ecosystem. The proposed approach is a compromise which relies on estimating
+actual compute resource demand based on an application metric (API request rate). Such estimate is inherently inaccurate,
+creating a need for a safety margin (excess replicas), and a risk of resource exhaustion under highly unusual application
+loads.
+
+#### Prometheus-metrics-adapter precludes the use of other sources of custom metrics
+With current K8s design, only one API extension can serve custom metrics. Per this GEP, the custom metrics extension
+point is occupied by prometheus-metrics-adapter.
+
+#### Heavyweight Prometheus infrastructure feeds signal back into cluster control loop
+The proposed signal flow brings two Prometheus instances (shoot control-plane prometheus and aggregate prometheus) into
+the shoot cluster control loop. This has major implications on reliability. The alternative is a small, custom-built
+component registered as seed custom metrics API extension, which directly calls the ShootKapi metrics endpoints and
+outputs request rate.
+
+The current approach was chosen solely because it leverages an existing open source component and avoids authoring a
+custom one. The historical record of Gardener Prometheus components reliability is deemed acceptable for the purpose.  
+
+The proposed complex signal path introduces additional latency to scaling response. At present, the resulting scaling
+latency is deemed satisfactory.
+
+While the scope of this GEP is limited to using the Prometheus signal path, the proposed design allows for an easy
+substitution with a signal path based on the aforementioned simple custom component. Such custom-built component,
+also has the advantage that it could easily coexist with other metrics providers by selectively forwarding queries to
+them. This supports future changes based on accumulated field experience, or evolving solution needs. 
+
+#### Observed fluctuations in actual compute efficiency affecting ShootKapi workloads
+There is a known source of horizontal scaling inefficiency, which presents an opportunity for substantial further
+improvement. It is not addressed by this proposal and instead deferred to a separate implementation stage, as it
+requires a strictly incremental effort on top of this GEP.
+
+An approximate 50% reduction (factor 0.5) of the effective compute power per unit of CPU was occasionally observed on
+some productive ShootKapi instances. The change occurs with a quick cutoff transition, and has been observed for periods
+or up to 2 hours. This variance currently forces substantial safety margin upon HPA's target metric value, effectively
+necessitating more replicas to be deployed than required during usual operation.
```

Comment:
I don't understand this section. What is the problem?

Reply:
Periods were observed, when kube-apiserver took twice the time to process the same requests, without a known change in relevant conditions. Cause needs to be understood, as this variation interferes with estimating how many 'requests per second' a Kapi instance can process.

Reply:
Marking the conversation "closed" to keep the PR manageable. Please reopen if necessary.